pop-api/examples/fungibles/lib.rs

```diff
@@ -1,108 +1,268 @@
+			api::allowance(self.id, owner, spender).unwrap_or_default()
 		}
 
+		/// Transfers `value` amount of tokens from the caller's account to account `to`
+		/// with additional `data` in unspecified format.
+		///
+		/// # Parameters
+		/// - `to` - The recipient account.
+		/// - `value` - The number of tokens to transfer.
+		/// - `data` - Additional data in unspecified format.
 		#[ink(message)]
-		pub fn transfer_from(
+		fn transfer(
 			&mut self,
-			token: TokenId,
-			from: AccountId,
 			to: AccountId,
 			value: Balance,
 			_data: Vec<u8>,
-		) -> Result<()> {
-			api::transfer_from(token, from, to, value)
+		) -> Result<(), Psp22Error> {
+			let caller = self.env().caller();
+			// No-op if the caller and `to` is the same address or `value` is zero.
+			if caller == to || value == 0 {
+				return Ok(());
+			}
+			api::transfer(self.id, to, value).map_err(Psp22Error::from)?;
+			self.env().emit_event(Transfer { from: Some(caller), to: Some(to), value });
+			Ok(())
 		}
 
+		/// Transfers `value` tokens on behalf of `from` to the account `to`
+		/// with additional `data` in unspecified format.
+		///
+		/// # Parameters
+		/// - `from` - The account from which the token balance will be withdrawn.
+		/// - `to` - The recipient account.
+		/// - `value` - The number of tokens to transfer.
+		/// - `data` - Additional data with unspecified format.
 		#[ink(message)]
-		pub fn approve(
+		fn transfer_from(
 			&mut self,
-			token: TokenId,
-			spender: AccountId,
+			from: AccountId,
+			to: AccountId,
 			value: Balance,
-		) -> Result<()> {
-			api::approve(token, spender, value)
+			_data: Vec<u8>,
+		) -> Result<(), Psp22Error> {
+			let caller = self.env().caller();
+			// No-op if `from` and `to` is the same address or `value` is zero.
+			if from == to || value == 0 {
+				return Ok(());
+			}
+			// If `from` and the caller are different addresses, a successful transfer results
+			// in decreased allowance by `from` to the caller and an `Approval` event with
+			// the new allowance amount is emitted.
+			api::transfer_from(self.id, from, to, value).map_err(Psp22Error::from)?;
+			self.env().emit_event(Transfer { from: Some(caller), to: Some(to), value });
+			self.env().emit_event(Approval {
+				owner: from,
+				spender: caller,
+				value: self.allowance(from, caller),
+			});
+			Ok(())
+		}
+
+		/// Approves `spender` to spend `value` amount of tokens on behalf of the caller.
+		///
+		/// Successive calls of this method overwrite previous values.
+		///
+		/// # Parameters
+		/// - `spender` - The account that is allowed to spend the tokens.
+		/// - `value` - The number of tokens to approve.
+		#[ink(message)]
+		fn approve(&mut self, spender: AccountId, value: Balance) -> Result<(), Psp22Error> {
+			let caller = self.env().caller();
+			// No-op if the caller and `spender` is the same address.
+			if caller == spender {
+				return Ok(());
+			}
+			api::approve(self.id, spender, value).map_err(Psp22Error::from)?;
```

Comment:
If the contract is the origin here, why would it have a function to allow anyone to call which allows the caller to spend the contracts funds?

I think this implementation just needs to return an error as its not possible to do this securely, as with increase/decrease allowance.

Reply:
I think we discussed about should we allow other accounts aside from contract to be the origin #301 and I don't know what is the final decision on this.